amr-wind/convection/incflo_godunov_minmod.H

```diff
@@ -0,0 +1,207 @@
+#ifndef GODUNOV_MINMOD_H
+#define GODUNOV_MINMOD_H
+
+#include <AMReX_Gpu.H>
+#include <AMReX_BCRec.H>
+
+/* This header file contains the inlined __host__ __device__ functions required
+   for the scalar advection routines for 3D Godunov. It also contains function
+   declarations for controlling host functions.  */
+
+namespace {
+
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void minmod(
+    const amrex::Real sm1,
+    const amrex::Real s0,
+    const amrex::Real sp1,
+    amrex::Real& dsm,
+    amrex::Real& dsp)
+{
+    // Calculate gradients on both sides
+    dsp = sp1 - s0;
+    dsm = s0 - sm1;
+
+    if (!(dsp * dsm < 0.0)) {
+        // Select the smaller slope if same sign
+        if (amrex::Math::abs(dsp) < amrex::Math::abs(dsm)) {
+            dsm = dsp;
+        } else {
+            dsp = dsm;
+        }
+    } else {
+        // Set to zero if opposite sign
+        dsp = 0.0;
+        dsm = 0.0;
+    }
+}
+
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void Godunov_minmod_bc(
+    const int n,
+    const amrex::Real sm1,
+    const amrex::Real s0,
+    const amrex::Real sp1,
+    amrex::Real& dsm,
+    amrex::Real& dsp,
+    const int bclo,
+    const int bchi,
+    const int domlo,
+    const int domhi)
+{
+    using namespace amrex;
+
+    if (bclo == BCType::ext_dir || bclo == BCType::hoextrap) {
+        if (n == domlo) {
+            // Ensure that left-side slope is used unchanged
+            dsm = s0 - sm1;
+        }
+    }
+
+    if (bchi == BCType::ext_dir || bchi == BCType::hoextrap) {
+        if (n == domhi) {
+            // Ensure that the right-side slope is used unchanged
+            dsp = sp1 - s0;
+        }
+    }
+}
+
+// This version is called after the MAC projection, when we use the
+// MAC-projected velocity
+//      for upwinding
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void Godunov_minmod_fpu_x(
+    const int i,
+    const int j,
+    const int k,
+    const int n,
+    const amrex::Real dt,
+    const amrex::Real dx,
+    amrex::Real& Im,
+    amrex::Real& Ip,
+    const amrex::Array4<const amrex::Real>& S,
+    const amrex::Array4<const amrex::Real>& vel_edge,
+    const amrex::BCRec& bc,
+    const int domlo,
+    const int domhi)
+{
+
+    using namespace amrex;
+
+    constexpr amrex::Real small_vel = 1e-10;
+
+    amrex::Real sm1 = S(i - 1, j, k, n);
+    amrex::Real s0 = S(i, j, k, n);
+    amrex::Real sp1 = S(i + 1, j, k, n);
+    amrex::Real dsp = 0.0;
+    amrex::Real dsm = 0.0;
+    minmod(sm1, s0, sp1, dsm, dsp);
+    Godunov_minmod_bc(
+        i, sm1, s0, sp1, dsm, dsp, bc.lo(0), bc.hi(0), domlo, domhi);
+
+    amrex::Real sigmap = amrex::Math::abs(vel_edge(i + 1, j, k)) * dt / dx;
+    amrex::Real sigmam = amrex::Math::abs(vel_edge(i, j, k)) * dt / dx;
+
+    if (vel_edge(i + 1, j, k) > small_vel) {
```

Comment:
This looks like an additional upwinding operator

Reply:
This isn't an additional upwinding operator, it's the pattern of assigning proper values to Ip and Im based on upwinding, which is present in similar routines like Godunov_ppm_fpu_x (line 647 of convection/incflo_godunov_ppm.H) and Godunov_weno_fpu_x (line 423 of convection/incflo_godunov_weno.H).